src/blosc2/ndarray.py

```diff
@@ -4575,60 +4585,189 @@ def matmul(x1: NDArray, x2: NDArray, **kwargs: Any) -> NDArray:
-
     n, k = x1.shape[-2:]
     m = x2.shape[-1]
+    result_shape = np.broadcast_shapes(x1.shape[:-2], x2.shape[:-2]) + (n, m)
+    result = blosc2.zeros(result_shape, dtype=np.result_type(x1, x2), **kwargs)
 
-    result = blosc2.zeros((n, m), dtype=np.result_type(x1, x2), **kwargs)
+    if 0 in result.shape + x1.shape + x2.shape:  # if any array is empty, return array of 0s
+        if x1_is_vector:
+            result.squeeze(axis=-2)
+        if x2_is_vector:
+            result.squeeze(axis=-1)
+        return result
 
     p, q = result.chunks[-2:]
     r = x2.chunks[-1]
 
-    for row in range(0, n, p):
-        row_end = builtins.min(row + p, n)
-        for col in range(0, m, q):
-            col_end = builtins.min(col + q, m)
-            for aux in range(0, k, r):
-                aux_end = builtins.min(aux + r, k)
-                bx1 = x1[row:row_end, aux:aux_end]
-                bx2 = x2[aux:aux_end, col:col_end]
-                result[row:row_end, col:col_end] += np.matmul(bx1, bx2)
+    intersecting_chunks = get_intersecting_chunks((), result.shape[:-2], result.chunks[:-2])
+    for chunk in intersecting_chunks:
+        chunk = chunk.raw
+        for row in range(0, n, p):
+            row_end = builtins.min(row + p, n)
+            for col in range(0, m, q):
+                col_end = builtins.min(col + q, m)
+                for aux in range(0, k, r):
+                    aux_end = builtins.min(aux + r, k)
+                    bx1 = (
+                        x1[chunk[-x1.ndim + 2 :] + (slice(row, row_end), slice(aux, aux_end))]
+                        if x1.ndim > 2
+                        else x1[row:row_end, aux:aux_end]
+                    )
+                    bx2 = (
+                        x2[chunk[-x2.ndim + 2 :] + (slice(aux, aux_end), slice(col, col_end))]
+                        if x2.ndim > 2
+                        else x2[aux:aux_end, col:col_end]
+                    )
+                    result[chunk + (slice(row, row_end), slice(col, col_end))] += np.matmul(bx1, bx2)
 
-    if x1_is_vector and x2_is_vector:
-        return result[0][0]
+    if x1_is_vector:
+        result.squeeze(axis=-2)
+    if x2_is_vector:
+        result.squeeze(axis=-1)
 
-    return result.squeeze()
+    return result
 
 
-def permute_dims(arr: NDArray, axes: tuple[int] | list[int] | None = None, **kwargs: Any) -> NDArray:
+def tensordot(
+    x1: NDArray, x2: NDArray, axes: int | tuple[Sequence[int], Sequence[int]] = 2, **kwargs: Any
+) -> NDArray:
+    """
+    Returns a tensor contraction of x1 and x2 over specific axes. The tensordot function corresponds to the generalized matrix product.
+
+    Parameters
+    ----------
+    x1: NDArray
+        First input array. Should have a numeric data type.
+
+    x2: NDArray
+        Second input array. Should have a numeric data type. Corresponding contracted axes of x1 and x2 must be equal.
+
+    axes: (Union[int, Tuple[Sequence[int], Sequence[int]]])
```

Comment:
Try to be consistent with the type declaration above. Also, I prefer syntax like `int | tuple[Sequence[int], Sequence[int]]` rather than more verbose `(Union[int, Tuple[Sequence[int], Sequence[int]]])`